lib/collection/src/shards/forward_proxy_shard.rs

```diff
@@ -210,31 +218,71 @@ impl ShardOperation for ForwardProxyShard {
         // the transfer needs to have access to the latest version of points.
         let mut result = self.wrapped_shard.update(operation.clone(), true).await?;
 
-        // Strip clock tag if forwarding to a different shard
-        // Each shard has their own clock tags and they are incompatible with other shards
-        if self.shard_id != self.remote_shard.id {
-            operation.clock_tag = None;
-        }
-
-        let remote_result = self
-            .remote_shard
-            .update(operation, false)
-            .await
-            .map_err(|err| CollectionError::forward_proxy_error(self.remote_shard.peer_id, err))?;
+        let forward_operation = if let Some(ring) = &self.resharding_hash_ring {
+            // If `ForwardProxyShard::resharding_hash_ring` is `Some`, we assume that proxy is used
+            // during *resharding* shard transfer, which forwards points to a remote shard with
+            // *different* shard ID.
+
+            debug_assert_ne!(self.shard_id, self.remote_shard.id);
+
+            // Only forward a *part* of the operation that belongs to remote shard.
+            let op = match operation.operation.split_by_shard(ring) {
+                OperationToShard::ToAll(op) => Some(op),
+                OperationToShard::ByShard(by_shard) => by_shard
+                    .into_iter()
+                    .find(|&(shard_id, _)| shard_id == self.remote_shard.id)
+                    .map(|(_, op)| op),
+            };
+
+            // Strip the clock tag from the operation, because clock tags are incompatible between
+            // different shards.
+            //
+            // Even though we expect (and assert) that this whole branch is only executed when
+            // forwarding to a *different* remote shard, we still handle the case when local and
+            // remote shards are the same, *just in case*.
+            //
+            // In such case `split_by_shard` call above would be a no-op, and we can preserve the
+            // clock tag.
+            let tag = if self.shard_id != self.remote_shard.id {
+                None
+            } else {
+                operation.clock_tag
+            };
+
+            op.map(|op| OperationWithClockTag::new(op, tag))
```

Comment:
I don't see a reason to keep the other branch, as we even assert we point to a different shard a few lines above.

```suggestion
            // Strip the clock tag from the operation, because clock tags are incompatible between
            // different shards.
            OperationWithClockTag::new(op, None)
```

Reply:
This is kinda intended.

If we ever accidentally pass `Some` hashring into a non-resharding forward proxy... it would still work as-is. Split by shard would just be a no-op in this case.

So Andrey wanted to structure this something like this:

```rust
let mut forward_operation = if let Some(ring) = hash_ring {
	// ...
} else {
    Some(operation)
};

if self.shard_id != self.remote_shard.id {
  forward_operation.as_mut().map(|op| op.operation.tag = None);
}
```

But, IMO, it makes `let Some(ring) = hash_ring` branch harder to understand, so I've moved `self.shard_id != self.remote_shard.id` under hashring branch.

I agree that it's kinda (like, 99%) pointless, but it is possible to miss `debug_assert`, and it does not hurt/degrade anything, so I don't mind having this.

I'll add a `log::warn` under that branch, though.